csharp/src/Ice/Communicator-LocatorManager.cs

```diff
@@ -74,19 +74,38 @@ public void Response(LocatorInfo locatorInfo, IObjectPrx? proxy)
                 }
             }
 
-            public void Exception(LocatorInfo locatorInfo, System.Exception exc)
+            public void Exception(Exception ex)
             {
-                try
-                {
-                    locatorInfo.GetEndpointsException(_ref, exc); // This throws.
-                }
-                catch (System.Exception ex)
+                Communicator communicator = _ref.GetCommunicator();
+                if (communicator.TraceLevels.Location > 0)
                 {
-                    if (_callback != null)
+                    if (ex is AdapterNotFoundException)
+                    {
+                        communicator.Logger.Trace(communicator.TraceLevels.LocationCat,
+                            $"adapter not found\nadapter = {_ref.GetAdapterId()}");
+                    }
+                    else if (ex is ObjectNotFoundException)
                     {
-                        _callback.SetException(ex);
+                        communicator.Logger.Trace(communicator.TraceLevels.LocationCat,
+                            $"object not found\nobject = {_ref.GetIdentity().ToString(communicator.ToStringMode)}");
+                    }
+                    else
+                    {
+                        var s = new System.Text.StringBuilder();
+                        s.Append("could not contact the locator to retrieve endpoints\n");
+                        if (_ref.GetAdapterId().Length > 0)
+                        {
+                            s.Append($"adapter = {_ref.GetAdapterId()}\n");
+                        }
+                        else
+                        {
+                            s.Append($"well-known proxy = {_ref}\n");
+                        }
+                        s.Append($"reason = {ex}");
+                        communicator.Logger.Trace(communicator.TraceLevels.LocationCat, s.ToString());
                     }
                 }
+                _callback?.SetException(ex);
```

Comment:
Shouldn't we change all the similar `if(_callback != null) { _callback.xxx() }` in this file as well?

Reply:
Yes, for a future PR :).